### PR TITLE
feat: implement all open issues (#9-#28)

### PR DESCRIPTION
**Pull request by Jurel89**
## Summary

Implements all 19 open GitHub issues across 4 waves, organized by dependency ordering:

### Wave 1 (commit 064758b)
- **#9** Fix year=0 identity check bug
- **#10** Pass normalized ISBN to sources in resolve_by_isbn
- **#11** Make BatchMatcher.link() truly streaming
- **#13** Document scoring philosophy for missing data
- **#14** Document hybrid_similarity discount constant
- **#17** Add publisher similarity scoring factor
- **#18** Add structured reason codes to MatchResult
- **#20** Add source diagnostics to BookResolver
- **#21** Improve normalize_authors to preserve structure (pairwise comparison)
- **#22** Wire SourceConfig into source implementations (mutually exclusive with #12)
- **#24** Add CLI tool for batch matching operations

### Wave 2 (commit fbb92c7)
- **#16** Add edge case tests for matching engine
- **#19** Add MatchKind enum for work-vs-edition classification
- **#27** Implement CONSENSUS resolve strategy
- **#28** Add series/volume scoring factor

### Wave 3
- **#15 + #23** Wire BatchConfig fields and implement parallel batch processing
- **#25** Create golden test dataset (50 curated book pairs) for regression testing

### Wave 4
- **#26** Add performance benchmark suite (`python -m benchmarks`)

## Test plan
- [x] All 234 tests pass (`pytest tests/ -q`)
- [x] Ruff lint clean (`ruff check src/ tests/ benchmarks/`)
- [x] Ruff format clean
- [x] mypy strict mode passes (`mypy src/`)
- [x] Golden dataset validates all 50 book pairs
- [x] Benchmarks run successfully (`python -m benchmarks`)
- [ ] CI pipeline passes

### CODE REVIEW

File: src/book_match/sources/resolver.py [hunk 1]
```diff
@@ -147,13 +158,138 @@ async def resolve(
...
+        # Find candidates that appear in enough distinct sources
+        consensus_candidates: set[int] = set()
+        for group in groups:
+            distinct_sources = {b.source for b in group if b.source}
+            if len(distinct_sources) >= self.min_agreeing_sources:
+                for b in group:
+                    consensus_candidates.add(id(b))
+
+        # Filter results to only consensus candidates
+        filtered = [r for r in results if id(r.remote_book) in consensus_candidates]
```
[the remark this answers is not in the document]
Comment (Jurel89) [reply src/book_match/sources/resolver.py]: Fixed in d3d6a4f. Now uses `set[Book]` since Book is a frozen dataclass and hashable.

File: src/book_match/sources/resolver.py [hunk 2]
```diff
@@ -147,13 +158,138 @@ async def resolve(
...
+
+        if not all_candidates:
+            return ResolveOutcome(results=[], source_diagnostics=tuple(diagnostics))
+
+        results = self.matcher.match_many(book, all_candidates, min_confidence=min_confidence)
+
+        return ResolveOutcome(
+            results=results[:max_results],
+            source_diagnostics=tuple(diagnostics),
+        )
```
[the remark this answers is not in the document]
Comment (Jurel89) [reply src/book_match/sources/resolver.py]: Fixed in d3d6a4f. `resolve_with_diagnostics()` now applies the configured `ResolveStrategy` (FIRST_CONFIDENT, ALL_SOURCES, CONSENSUS) matching the behavior of `resolve()`.

File: src/book_match/batch/processor.py [hunk 3]
```diff
@@ -100,43 +102,42 @@ def deduplicate(
...
-                    if pair in compared:
-                        continue
-                    compared.add(pair)
-
-                    # Compare
-                    book1 = books[idx1]
-                    book2 = books[idx2]
+                    if pair not in seen:
+                        seen.add(pair)
+                        all_pairs.append(pair)
```
[the remark this answers is not in the document]
Comment (Jurel89) [reply src/book_match/batch/processor.py]: The previous implementation also stored O(k) data via the `compared: set[tuple[int, int]]` tracking set, so peak memory is comparable. For the parallel path, pairs must be materialized to chunk them for the thread pool. For the single-threaded path, we'd still need a `seen` set of the same size. The net memory difference is marginal — both approaches are O(unique_pairs). If this becomes a real bottleneck for very large catalogs, a streaming chunk iterator could be added as a follow-up.

File: src/book_match/matching/engine.py [hunk 4]
```diff
@@ -368,9 +512,28 @@ def match(self, local: Book, remote: Book) -> MatchResult:
 
         factors.extend([title_factor, author_factor, year_factor, language_factor])
 
+        # Optional publisher factor (only when weight > 0)
+        if self.config.publisher_weight > 0:
+            publisher_factor = self._compare_publishers(local.publisher, remote.publisher)
+            factors.append(publisher_factor)
```
[the remark this answers is not in the document]
Comment (Jurel89) [reply src/book_match/matching/engine.py]: Fixed in f82078e. Publisher factor is now always computed (with weight=0.0 contribution when disabled), so MatchKind classification has consistent access to publisher similarity regardless of config.

File: src/book_match/sources/resolver.py [hunk 5]
```diff
@@ -59,13 +69,15 @@ def __init__(
...
             match_config: Configuration for matching (ignored if matcher provided)
             strategy: Resolution strategy
+            min_agreeing_sources: Minimum sources that must agree for CONSENSUS strategy
         """
         if not sources:
             raise ValueError("At least one metadata source is required")
 
         self.sources = list(sources)
         self.matcher = matcher or BookMatcher(match_config)
         self.strategy = strategy
```
[the remark this answers is not in the document]
Comment (Jurel89) [reply src/book_match/sources/resolver.py]: Fixed in f82078e. Added validation: `min_agreeing_sources >= 2` and `<= len(sources)` when using CONSENSUS strategy, raising `ValueError` otherwise.
